libraries/AP_Mount/AP_Mount_Siyi.h

```diff
@@ -0,0 +1,192 @@
+    void process_packet();
+
+    // send packet to gimbal
+    void send_packet(SiyiCommandId cmd_id, const uint8_t* databuff, uint8_t databuff_len);
+    void send_1byte_packet(SiyiCommandId cmd_id, uint8_t data_byte);
+
+    // request info from gimbal
+    void request_firmware_version() { send_packet(SiyiCommandId::ACQUIRE_FIRMWARE_VERSION, nullptr, 0); }
+    void request_hardware_id() { send_packet(SiyiCommandId::HARDWARE_ID, nullptr, 0); }
+    void request_configuration() { send_packet(SiyiCommandId::ACQUIRE_GIMBAL_CONFIG_INFO, nullptr, 0); }
+    void request_function_feedback_info() { send_packet(SiyiCommandId::FUNCTION_FEEDBACK_INFO, nullptr, 0); }
+    void request_gimbal_attitude() { send_packet(SiyiCommandId::ACQUIRE_GIMBAL_ATTITUDE, nullptr, 0); }
+
+    // rotate gimbal.  pitch_rate and yaw_rate are scalars in the range -100 ~ +100
+    // yaw_is_ef should be true if gimbal should maintain an earth-frame target (aka lock)
+    void rotate_gimbal(int8_t pitch_scalar, int8_t yaw_scalar, bool yaw_is_ef);
+
+    // center gimbal
+    void center_gimbal();
+
+    // set gimbal's lock vs follow mode
+    // lock should be true if gimbal should maintain an earth-frame target
+    // lock is false to follow / maintain a body-frame target
+    void set_lock(bool lock);
+
+    // send target pitch and yaw rates to gimbal
+    // yaw_is_ef should be true if yaw_rads target is an earth frame rate, false if body_frame
+    void send_target_rates(float pitch_rads, float yaw_rads, bool yaw_is_ef);
+
+    // send target pitch and yaw angles to gimbal
+    // yaw_is_ef should be true if yaw_rad target is an earth frame angle, false if body_frame
+    void send_target_angles(float pitch_rad, float yaw_rad, bool yaw_is_ef);
+
+    // internal variables
+    AP_HAL::UARTDriver *_uart;                      // uart connected to gimbal
+    bool _initialised;                              // true once the driver has been initialised
+    bool _got_firmware_version;                     // true once gimbal firmware version has been received
+
+    // buffer holding bytes from latest packet.  This is only used to calculate the crc
+    uint8_t _msg_buff[AP_MOUNT_SIYI_PACKETLEN_MAX];
+    uint8_t _msg_buff_len;
+    const uint8_t _msg_buff_data_start = 8;         // data starts at this byte of _msg_buff
+
+    // the order of fields matches the incoming message so it can be used to calculate the crc
+    struct PACKED {
+        uint8_t ctrl;                               // control byte. bit 0: need_ack, bit 1: this is ack_packet
+        uint16_t data_len;                          // expected number of data bytes
+        uint16_t seq;                               // sequence number for this packet
+        uint8_t command_id;                         // command id
+        uint16_t data_bytes_received;               // number of data bytes received so far
+        uint16_t crc16;                             // latest message's crc
```

Comment:
Is this actually used?

Most of these aren't actually relevant - you can retrieve them from the _message_buf once the checksum has passed (and when you need them), rather than stashing them away as the parser moves through its states.  That's probably too big a change for *right now*.  

As an aside, and I'm not suggesting changing a mostly-working driver, `SIM_PS_Lightware_SF45B.cpp` parses a very similar protocol to this one.  Once the message in the receive buffer is checksummed it can reference fields from each different message type by name (see `handle_message` over there),  It also uses the same constructs to send packets of a specific type.  It also has better resync semantics as it doesn't throw away any data unnecessarily...

Reply:
Txs, I've removed the seq and ctrl fields.  As mentioned above during development I changed the way the crc was calculated so that it runs once on the complete message instead of on each individual field which led to some duplication.